tests/lax_test.py

```diff
@@ -1714,26 +1714,42 @@ def testTransposeAgainstNumpy(self, shape, dtype, perm):
       {"testcase_name": "_op={}_inshape={}_reducedims={}_initval={}"
        .format(op.__name__, jtu.format_shape_dtype_string(shape, dtype), dims,
                init_val),
-       "op": op, "init_val": init_val, "shape": shape, "dtype": dtype, "dims": dims}
-      for init_val, op, types in [
-          (0, lax.add, default_dtypes),
-          (1, lax.mul, default_dtypes),
-          (0, lax.max, all_dtypes), # non-monoidal
-          (-np.inf, lax.max, float_dtypes),
-          (dtypes.iinfo(np.int32).min, lax.max, [np.int32]),
-          (dtypes.iinfo(np.int64).min, lax.max, [np.int64]),
-          (np.inf, lax.min, float_dtypes),
-          (dtypes.iinfo(np.int32).max, lax.min, [np.int32]),
-          (dtypes.iinfo(np.int64).max, lax.min, [np.int64]),
-          (dtypes.iinfo(np.uint32).max, lax.min, [np.uint32]),
-          (dtypes.iinfo(np.uint64).max, lax.min, [np.uint64]),
+       "op": op, "init_val": init_val, "shape": shape, "dtype": dtype, "dims": dims, "prim": prim}
+      for init_val, op, types, prim in [
+          (0, lax.add, default_dtypes, lax.reduce_sum_p),
+          (1, lax.mul, default_dtypes, lax.reduce_prod_p),
+          (0, lax.max, float_dtypes + complex_dtypes + int_dtypes, lax.reduce_p), # non-monoidal
+          (0, lax.max, uint_dtypes + bool_dtypes, lax.reduce_max_p),
+          (-np.inf, lax.max, float_dtypes, lax.reduce_max_p),
+          (dtypes.iinfo(np.int32).min, lax.max, [np.int32], lax.reduce_max_p),
+          (dtypes.iinfo(np.int64).min, lax.max, [np.int64], lax.reduce_max_p),
+          (np.inf, lax.min, float_dtypes, lax.reduce_min_p),
+          (dtypes.iinfo(np.int32).max, lax.min, [np.int32], lax.reduce_min_p),
+          (dtypes.iinfo(np.int64).max, lax.min, [np.int64], lax.reduce_min_p),
+          (dtypes.iinfo(np.uint32).max, lax.min, [np.uint32], lax.reduce_min_p),
+          (dtypes.iinfo(np.uint64).max, lax.min, [np.uint64], lax.reduce_min_p),
+          (dtypes.iinfo(np.uint32).max, lax.bitwise_and, [np.uint32], lax.reduce_and_p),
+          (dtypes.iinfo(np.uint64).min, lax.bitwise_or, [np.uint64], lax.reduce_or_p),
+          (0, lax.bitwise_xor, uint_dtypes, lax.reduce_xor_p)
       ]
       for dtype in types
       for shape, dims in [
           [(3, 4, 5), (0,)], [(3, 4, 5), (1, 2)],
           [(3, 4, 5), (0, 2)], [(3, 4, 5), (0, 1, 2)]
       ]))
-  def testReduce(self, op, init_val, shape, dtype, dims):
+  def testReduce(self, op, init_val, shape, dtype, dims, prim):
+    def reference_fun(operand, init_value):
+      np_op_name = op.__name__
+      if op.__name__ == "mul":
+        np_op_name = "multiply"
+
+      np_op = getattr(np, np_op_name)
+
+      if hasattr(np_op, "reduce"):
+        return np_op.reduce(operand, axis=dims, initial=init_value)
+      else:
+        return np_op(operand, axis=dims)
```

Comment:
I'm a bit concerned about how this test is set up – there is some amount of "action at a distance" that makes the parameterization feel a bit brittle – for example, this reference implementation doesn't always reference the `init_val`, which means there's additional, unmentioned requirements on the input parameterization for the test to be correct. It also makes it hard to read through the test and understand what's going on.

I'd suggest using a class-record based test parameterization and adding the appropriate numpy function to those entries (I think since `init_val` is the obvious choice in each case, you don't even need to pass it to the reference function)

Reply:
Done. The initial value is not part of the reference function signature anymore, and the parametrization has been changed to iterate over the newly created `ReducerOpRecord`s.